endpoints/openrtb2/firstpartydata/first_party_data.go

```diff
@@ -0,0 +1,238 @@
+package firstpartydata
+
+import (
+	"encoding/json"
+	"errors"
+	"github.com/buger/jsonparser"
+	"github.com/evanphx/json-patch"
+	"github.com/mxmCherry/openrtb/v15/openrtb2"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"github.com/prebid/prebid-server/util/jsonutil"
+)
+
+const (
+	site = "site"
+	app  = "app"
+	user = "user"
+	data = "data"
+)
+
+func GetFPDData(request []byte) ([]byte, map[string][]byte, error) {
+	//If {site,app,user}.data exists, merge it into {site,app,user}.ext.data and remove {site,app,user}.data
+
+	fpdReqData := make(map[string][]byte, 0)
+	request, siteFPD, err := jsonutil.FindAndDropElement(request, site, data)
+	if err != nil {
+		return request, nil, err
+	}
+	fpdReqData[site] = siteFPD
+
+	request, appFPD, err := jsonutil.FindAndDropElement(request, app, data)
+	if err != nil {
+		return request, nil, err
+	}
+	fpdReqData[app] = appFPD
+
+	fpdReqData[user] = []byte{}
+	userDataBytes, _, _, err := jsonparser.Get(request, user, data)
+	if err != nil && err != jsonparser.KeyPathNotFoundError {
+		return request, nil, err
+	}
+
+	if len(userDataBytes) > 0 {
+		var userData []openrtb2.Data
+		userDataCopy := make([]byte, len(userDataBytes))
+		copy(userDataCopy, userDataBytes)
+		err = json.Unmarshal(userDataCopy, &userData)
+		if err != nil {
+			//unable to unmarshal to []openrtb2.Data, meaning this is FPD data
+			request, err = jsonutil.DropElement(request, user, data)
+			if err != nil {
+				return request, nil, err
+			}
+			fpdReqData[user] = userDataCopy
+		}
+	}
+
+	return request, fpdReqData, nil
+}
+
+func BuildFPD(bidRequest *openrtb2.BidRequest, fpdBidderData map[openrtb_ext.BidderName]*openrtb_ext.FPDData, firstPartyData map[string][]byte) (map[openrtb_ext.BidderName]*openrtb_ext.FPDData, []error) {
+
+	// If an attribute doesn't pass defined validation checks,
+	// it should be removed from the request with a warning placed
+	// in the messages section of debug output
+	// The auction should continue
+
+	errL := make([]error, 0)
+	resolvedFpdData := make(map[openrtb_ext.BidderName]*openrtb_ext.FPDData)
+
+	for bidderName, fpdConfig := range fpdBidderData {
+
+		resolvedFpdConfig := &openrtb_ext.FPDData{}
+
+		if fpdConfig.User != nil {
+			if bidRequest.User == nil {
+				resolvedFpdConfig.User = fpdConfig.User
+			} else {
```

Comment:
If the User logic here was in its own function, this code could be flattened by using a couple of early returns instead of the nested `if` statements and the `else` could be eliminated which would improve readability and testability IMO.

Reply:
Refactored